codec-http2/src/main/java/io/netty/handler/codec/http2/AbstractHttp2StreamChannel.java

```diff
@@ -168,72 +210,54 @@ protected final void doWrite(ChannelOutboundBuffer in) throws Exception {
         if (closed) {
             throw CLOSED_CHANNEL_EXCEPTION;
         }
-
-        EventExecutor preferredExecutor = preferredEventExecutor();
-
-        // TODO: this is pretty broken; futures should only be completed after they are processed on
-        // the parent channel. However, it isn't currently possible due to ChannelOutboundBuffer's
-        // behavior which requires completing the current future before getting the next message. It
-        // should become easier once we have outbound flow control support.
-        // https://github.com/netty/netty/issues/4941
-        if (preferredExecutor.inEventLoop()) {
-            for (;;) {
-                Object msg = in.current();
-                if (msg == null) {
-                    break;
-                }
-                try {
-                    doWrite(ReferenceCountUtil.retain(msg));
-                } catch (Throwable t) {
-                    // It would be nice to fail the future, but we can't do that if not on the event
-                    // loop. So we instead opt for a solution that is consistent.
-                    pipeline().fireExceptionCaught(t);
-                }
-                in.remove();
+        final MessageSizeEstimator.Handle sizeEstimator = config().getMessageSizeEstimator().newHandle();
+        for (;;) {
+            final Object msg = in.current();
+            if (msg == null) {
+                break;
             }
-            doWriteComplete();
-        } else {
-            // Use a copy because the original msgs will be recycled by AbstractChannel.
-            final Object[] msgsCopy = new Object[in.size()];
-            for (int i = 0; i < msgsCopy.length; i ++) {
-                msgsCopy[i] = ReferenceCountUtil.retain(in.current());
+            // TODO(buchgr): Detecting cancellation relies on ChannelOutboundBuffer internals. NOT COOL!
+            if (msg == Unpooled.EMPTY_BUFFER /* The write was cancelled. */
+                || msg == REEVALUATE_WRITABILITY_MESSAGE /* Write to trigger writability after window update. */) {
                 in.remove();
+                continue;
+            }
+            final int bytes = sizeEstimator.size(msg);
+            /**
+             * The flow control window needs to be decrement before stealing the message from the buffer (and thereby
+             * decrementing the number of pending bytes). Else, when calling steal() the number of pending bytes could
+             * be less than the writebuffer watermark (=flow control window) and thus trigger a writability change.
+             *
+             * This code must never trigger a writability change. Only reading window updates or channel writes may
+             * change the channel's writability.
+             */
+            incrementOutboundFlowControlWindow(-bytes);
+            final ChannelPromise promise = in.steal();
```

Comment:
The addition of steal() and its use here fixes a bug (#4941) that we're running into in Finagle land. Since it seems like a reasonably compartmentalized piece, is it possible to split out the steal() fix into a separate PR?